File: apps/downloadx/tests/unit/chunkScheduler.test.ts
import { describe, expect, it } from 'vitest';

import type { Chunk } from '../../src/chunk.js';
import { findSplitCandidate, planChunks } from '../../src/chunkScheduler.js';

describe('planChunks', () => {
  it('returns a single empty chunk for zero-size files', () => {
    const plans = planChunks({ totalSize: 0, targetChunkCount: 4, minChunkSize: 16 });
    expect(plans).toEqual([{ offset: 0, length: 0, downloadedBytes: 0 }]);
  });

  it('falls back to one chunk when total is below minChunkSize', () => {
    const plans = planChunks({ totalSize: 10, targetChunkCount: 4, minChunkSize: 64 });
    expect(plans).toEqual([{ offset: 0, length: 10, downloadedBytes: 0 }]);
  });

  it('divides into targetChunkCount pieces with the tail absorbing remainder', () => {
    const plans = planChunks({ totalSize: 1003, targetChunkCount: 4, minChunkSize: 16 });
    expect(plans).toHaveLength(4);
    const total = plans.reduce((acc, p) => acc + p.length, 0);
    expect(total).toBe(1003);
    expect(plans[0]?.offset).toBe(0);
    expect(plans[1]?.offset).toBe(plans[0]!.length);
  });

  it('caps chunk count so no chunk is below minChunkSize', () => {
    const plans = planChunks({ totalSize: 100, targetChunkCount: 8, minChunkSize: 30 });
    // floor(100/30) = 3 → at most three chunks.
    expect(plans).toHaveLength(3);
  });

  it('returns the resumeFrom plan verbatim when provided', () => {
    const prior = [
      { offset: 0, length: 100, downloadedBytes: 50 },
      { offset: 100, length: 100, downloadedBytes: 0 },
    ];
    const plans = planChunks({
      totalSize: 200,
      targetChunkCount: 4,
      minChunkSize: 16,
      resumeFrom: prior,
    });
    expect(plans).toEqual(prior);
    // returned plans are independent copies
    expect(plans).not.toBe(prior);
  });
});

interface FakeChunk {
  status: string;
  remainingBytes: number;
  quality: string;
  truncated?: { offset: number; length: number } | null;
}

function makeFake(over: Partial<FakeChunk>): Chunk {
  const base: FakeChunk = {
    status: 'downloading',
    remainingBytes: 1024,
    quality: 'good',
    ...over,
  };
  const truncateTail = (_min: number): { offset: number; length: number } | null => {
    if (base.truncated !== undefined) return base.truncated;
    // Default behaviour: split exactly in half from the tail.
    const len = Math.floor(base.remainingBytes / 2);
    const removed = { offset: 10_000, length: len };
    base.remainingBytes -= len;
    return removed;
  };
  return {
    status: base.status,
    remainingBytes: base.remainingBytes,
    quality: base.quality,
    truncateTail,
  } as unknown as Chunk;
}

describe('findSplitCandidate', () => {
  it('returns null when the pool is full', () => {
    const chunks = [makeFake({}), makeFake({})];
    const r = findSplitCandidate({
      activeChunks: chunks,
      maxChunks: 2,
      minChunkSize: 100,
      trigger: 'completed-reassign',
    });
    expect(r).toBeNull();
  });

  it('ignores chunks that are not downloading', () => {
    const chunks = [makeFake({ status: 'paused' })];
    const r = findSplitCandidate({
      activeChunks: chunks,
      maxChunks: 4,
      minChunkSize: 100,
      trigger: 'completed-reassign',
    });
    expect(r).toBeNull();
  });

  it('ignores chunks with not enough remaining bytes (< 2x minChunkSize)', () => {
    const chunks = [makeFake({ remainingBytes: 150 })];
    const r = findSplitCandidate({
      activeChunks: chunks,
      maxChunks: 4,
      minChunkSize: 100,
      trigger: 'completed-reassign',
    });
    expect(r).toBeNull();
  });

  it('prefers stalled over poor over good; ties broken by remaining size', () => {
    const good = makeFake({ remainingBytes: 10_000, quality: 'good' });
    const poor = makeFake({ remainingBytes: 5_000, quality: 'poor' });
    const stalled = makeFake({ remainingBytes: 1_000, quality: 'stalled' });
    const r = findSplitCandidate({
      activeChunks: [good, poor, stalled],
      maxChunks: 10,
      minChunkSize: 100,
      trigger: 'completed-reassign',
    });
    expect(r?.chunk).toBe(stalled);
  });

  it('propagates the trigger reason to the returned candidate', () => {
    const c = makeFake({});
    const r = findSplitCandidate({
      activeChunks: [c],
      maxChunks: 10,
      minChunkSize: 100,
      trigger: 'slow',
    });
    expect(r?.reason).toBe('slow');
  });
});

File: apps/downloadx_dart/test/unit/chunk_scheduler_test.dart
import 'package:downloadx/downloadx.dart';
import 'package:test/test.dart';

void main() {
  group('planChunks', () {
    test('single chunk when smaller than minChunkSize', () {
      final plans = planChunks(const PlanOptions(
        totalSize: 500,
        targetChunkCount: 4,
        minChunkSize: 1000,
      ));
      expect(plans.length, 1);
      expect(plans.first.length, 500);
    });

    test('single chunk when targetChunkCount <= 1', () {
      final plans = planChunks(const PlanOptions(
        totalSize: 10000,
        targetChunkCount: 1,
        minChunkSize: 10,
      ));
      expect(plans.length, 1);
      expect(plans.first.length, 10000);
    });

    test('tiles the file with no gaps or overlaps', () {
      final plans = planChunks(const PlanOptions(
        totalSize: 1000,
        targetChunkCount: 4,
        minChunkSize: 100,
      ));
      expect(plans.length, 4);
      var expectedOffset = 0;
      var sum = 0;
      for (final p in plans) {
        expect(p.offset, expectedOffset);
        expectedOffset += p.length;
        sum += p.length;
      }
      expect(sum, 1000);
      expect(plans.last.offset + plans.last.length, 1000);
    });

    test('count is bounded by totalSize / minChunkSize', () {
      final plans = planChunks(const PlanOptions(
        totalSize: 1000,
        targetChunkCount: 100,
        minChunkSize: 250,
      ));
      // 1000 / 250 = 4 max chunks
      expect(plans.length, 4);
    });

    test('preserves resumeFrom layout', () {
      final plans = planChunks(PlanOptions(
        totalSize: 1000,
        targetChunkCount: 4,
        minChunkSize: 100,
        resumeFrom: const [
          ChunkPlan(offset: 0, length: 600, downloadedBytes: 300),
          ChunkPlan(offset: 600, length: 400, downloadedBytes: 0),
        ],
      ));
      expect(plans.length, 2);
      expect(plans.first.downloadedBytes, 300);
    });
  });
}
